sam0_common: Make SPI peripheral DMA compatible

bergzand opened the pull request:
### Contribution description

This adds DMA compatibility to the SPI peripheral on the atsam platform. DMA can be enabled on a per peripheral basis. As the unconfigured DMA option is value zero, leaving the DMA unconfigured with a peripheral while having DMA itself enabled doesn't break the peripheral. (currently the case on the stm32 :cry:). 

A SPI peripheral will claim two DMA channels without releasing them. This prevents having to reconfigure the DMA channel on every transfer and keeps the transfer overhead as low as possible. The receive channel has a higher priority than the transfer channel to keep the transfer channel from starving the receive channel.

SPI register writes are combined in two DMA chained descriptors. This saves considerable time by combining the two transfers into a single blocking transfer.

### Testing procedure

As described in #14260, use `tests/periph_spi` with `periph_dma` in the `FEATURES_REQUIRED`. The benchmark command must work at 10Mhz. It should be tested on:
 - [x] same54-xpro
 - [x] samr21-xpro
 - [x] samr30-xpro/saml21-xpro
 - [x] saml11-xpro/saml10-xpro

### Benchmarks

#### Benchmarks on the samr21-xpro with the SPI peripheral configured at 10Mhz:

<details>
 <summary>With DMA</summary>

```
init 1 0 4 1 3
2020-06-11 11:35:36,424 #  init 1 0 4 1 3
2020-06-11 11:35:36,429 # SPI_DEV(1) initialized: mode: 0, clk: 4, cs_port: 1, cs_pin: 3
bench 3
2020-06-11 11:35:38,088 #  bench
2020-06-11 11:35:38,093 # ### Running some benchmarks, all values in [us] ###
2020-06-11 11:35:38,096 # ### Test                              Transfer time   user time
2020-06-11 11:35:38,097 # 
2020-06-11 11:35:38,122 #  1 - write 1000 times 1 byte:                 22196   22215
2020-06-11 11:35:38,147 #  2 - write 1000 times 2 byte:                 21738   21757
2020-06-11 11:35:38,261 #  3 - write 1000 times 100 byte:               109009  28027
2020-06-11 11:35:38,291 #  4 - write 1000 times 1 byte to register:     25759   25778
2020-06-11 11:35:38,321 #  5 - write 1000 times 2 byte to register:     25217   25236
2020-06-11 11:35:38,440 #  6 - write 1000 times 100 byte to register:   113009  31028
2020-06-11 11:35:38,466 #  7 - read 1000 times 2 byte:                  21884   21903
2020-06-11 11:35:38,578 #  8 - read 1000 times 100 byte:                109009  28027
2020-06-11 11:35:38,609 #  9 - read 1000 times 2 byte from register:    25363   25382
2020-06-11 11:35:38,727 # 10 - read 1000 times 100 byte from register:  113010  31028
2020-06-11 11:35:38,753 # 11 - transfer 1000 times 2 byte:              21800   21819
2020-06-11 11:35:38,867 # 12 - transfer 1000 times 100 byte:            109009  28027
2020-06-11 11:35:38,897 # 13 - transfer 1000 times 2 byte to register:  25300   25319
2020-06-11 11:35:39,016 # 14 - transfer 1000 times 100 byte to register:113338  31028
2020-06-11 11:35:39,030 # 15 - acquire/release 1000 times:              10717   10736
2020-06-11 11:35:40,033 # -- - SUM:                                     866358  377310
2020-06-11 11:35:40,034 # 
2020-06-11 11:35:40,036 # ### All runs complete ###
```
</details>
<details>
 <summary>Without DMA</summary>

```
init 1 0 4 1 3
2020-06-11 11:40:28,612 #  init 1 0 4 1 3
2020-06-11 11:40:28,617 # SPI_DEV(1) initialized: mode: 0, clk: 4, cs_port: 1, cs_pin: 3
bench 3
2020-06-11 11:40:29,736 #  bench
2020-06-11 11:40:29,740 # ### Running some benchmarks, all values in [us] ###
2020-06-11 11:40:29,743 # ### Test                              Transfer time   user time
2020-06-11 11:40:29,743 # 
2020-06-11 11:40:29,751 #  1 - write 1000 times 1 byte:                 4342    4361
2020-06-11 11:40:29,761 #  2 - write 1000 times 2 byte:                 5467    5486
2020-06-11 11:40:29,922 #  3 - write 1000 times 100 byte:               156509  156528
2020-06-11 11:40:29,934 #  4 - write 1000 times 1 byte to register:     8301    8320
2020-06-11 11:40:29,949 #  5 - write 1000 times 2 byte to register:     9488    9507
2020-06-11 11:40:30,115 #  6 - write 1000 times 100 byte to register:   160571  160590
2020-06-11 11:40:30,124 #  7 - read 1000 times 2 byte:                  5551    5570
2020-06-11 11:40:30,285 #  8 - read 1000 times 100 byte:                156675  156694
2020-06-11 11:40:30,299 #  9 - read 1000 times 2 byte from register:    9592    9611
2020-06-11 11:40:30,465 # 10 - read 1000 times 100 byte from register:  160717  160736
2020-06-11 11:40:30,475 # 11 - transfer 1000 times 2 byte:              5530    5549
2020-06-11 11:40:30,638 # 12 - transfer 1000 times 100 byte:            158696  158715
2020-06-11 11:40:30,652 # 13 - transfer 1000 times 2 byte to register:  9509    9528
2020-06-11 11:40:30,821 # 14 - transfer 1000 times 100 byte to register:162738  162757
2020-06-11 11:40:30,835 # 15 - acquire/release 1000 times:              10676   10695
2020-06-11 11:40:31,839 # -- - SUM:                                     1024362 1024647
2020-06-11 11:40:31,839 # 
2020-06-11 11:40:31,842 # ### All runs complete ###
```
</details>

#### Benchmarks on the samr21-xpro with the SPI peripheral configured at 1Mhz:
<details>
 <summary>With DMA</summary>

```
init 1 0 2 1 3
2020-06-11 11:38:18,898 #  init 1 0 2 1 3
2020-06-11 11:38:18,904 # SPI_DEV(1) initialized: mode: 0, clk: 2, cs_port: 1, cs_pin: 3
> bench
2020-06-11 11:38:20,449 #  bench
2020-06-11 11:38:20,454 # ### Running some benchmarks, all values in [us] ###
2020-06-11 11:38:20,457 # ### Test                              Transfer time   user time
2020-06-11 11:38:20,457 # 
2020-06-11 11:38:20,508 #  1 - write 1000 times 1 byte:                 47009   28028
2020-06-11 11:38:20,559 #  2 - write 1000 times 2 byte:                 47008   28027
2020-06-11 11:38:21,395 #  3 - write 1000 times 100 byte:               831006  28027
2020-06-11 11:38:21,451 #  4 - write 1000 times 1 byte to register:     51008   32027
2020-06-11 11:38:21,513 #  5 - write 1000 times 2 byte to register:     58009   31028
2020-06-11 11:38:22,361 #  6 - write 1000 times 100 byte to register:   842009  31028
2020-06-11 11:38:22,412 #  7 - read 1000 times 2 byte:                  47008   28027
2020-06-11 11:38:23,247 #  8 - read 1000 times 100 byte:                831008  28027
2020-06-11 11:38:23,310 #  9 - read 1000 times 2 byte from register:    58009   31028
2020-06-11 11:38:24,158 # 10 - read 1000 times 100 byte from register:  842009  31028
2020-06-11 11:38:24,209 # 11 - transfer 1000 times 2 byte:              47008   28027
2020-06-11 11:38:25,045 # 12 - transfer 1000 times 100 byte:            831008  28027
2020-06-11 11:38:25,108 # 13 - transfer 1000 times 2 byte to register:  58009   31028
2020-06-11 11:38:25,956 # 14 - transfer 1000 times 100 byte to register:842009  31028
2020-06-11 11:38:25,971 # 15 - acquire/release 1000 times:              11363   11382
2020-06-11 11:38:26,975 # -- - SUM:                                     5443480 425767
2020-06-11 11:38:26,976 # 
2020-06-11 11:38:26,978 # ### All runs complete ###
```
</details>
<details>
 <summary>Without DMA</summary>

```
init 1 0 2 1 3
2020-06-11 11:39:36,725 #  init 1 0 2 1 3
2020-06-11 11:39:36,731 # SPI_DEV(1) initialized: mode: 0, clk: 2, cs_port: 1, cs_pin: 3
bench 3
2020-06-11 11:39:37,603 #  bench
2020-06-11 11:39:37,607 # ### Running some benchmarks, all values in [us] ###
2020-06-11 11:39:37,610 # ### Test                              Transfer time   user time
2020-06-11 11:39:37,610 # 
2020-06-11 11:39:37,626 #  1 - write 1000 times 1 byte:                 11780   11799
2020-06-11 11:39:37,651 #  2 - write 1000 times 2 byte:                 20342   20361
2020-06-11 11:39:38,556 #  3 - write 1000 times 100 byte:               900259  900278
2020-06-11 11:39:38,583 #  4 - write 1000 times 1 byte to register:     23175   23194
2020-06-11 11:39:38,620 #  5 - write 1000 times 2 byte to register:     31801   31820
2020-06-11 11:39:39,537 #  6 - write 1000 times 100 byte to register:   911759  911778
2020-06-11 11:39:39,562 #  7 - read 1000 times 2 byte:                  20425   20444
2020-06-11 11:39:40,467 #  8 - read 1000 times 100 byte:                900425  900444
2020-06-11 11:39:40,504 #  9 - read 1000 times 2 byte from register:    31905   31924
2020-06-11 11:39:41,421 # 10 - read 1000 times 100 byte from register:  911905  911924
2020-06-11 11:39:41,446 # 11 - transfer 1000 times 2 byte:              20405   20424
2020-06-11 11:39:42,353 # 12 - transfer 1000 times 100 byte:            902446  902465
2020-06-11 11:39:42,390 # 13 - transfer 1000 times 2 byte to register:  31821   31840
2020-06-11 11:39:43,310 # 14 - transfer 1000 times 100 byte to register:913925  913944
2020-06-11 11:39:43,325 # 15 - acquire/release 1000 times:              11342   11361
2020-06-11 11:39:44,329 # -- - SUM:                                     5643715 5644000
2020-06-11 11:39:44,329 # 
2020-06-11 11:39:44,331 # ### All runs complete ###
```
</details>

Visible is that the overhead of enabling DMA is relative constant, especially at lower SPI clock frequencies. With register writes the peripheral makes use of DMA transfer chaining, making the extra SPI write low overhead. Another notable result is that with large transfers the DMA based config is overal faster than without DMA. At some point, depending on the SPI clock speed, it pays of to always use DMA for the transfer.

### Issues/PRs references

Depends on #14260

bergzand commented:
@MrKevinWeiss Do you have time somewhere these weeks to test this on the HIL platform?

benpicco commented:
I see that the transfer time greatly increases for small (few byte) transfers - is it possible to only start a DMA transfer if the transfer size is greater than a certain threshold and do the old polling otherwise?

benpicco commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -164,30 +194,116 @@ void spi_release(spi_t bus)
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t)cs);
+    }
+    if (_use_dma(bus)) {
```
```suggestion
    if (_use_dma(bus) && len > DMA_THRESHOLD) {
```

bergzand commented:
> I see that the transfer time greatly increases for small (few byte) transfers - is it possible to only start a DMA transfer if the transfer size is greater than a certain threshold and do the old polling otherwise?

As DMA usage is fully optional at the moment and won't affect our default examples, I'm going to leave this out-of-scope for now. Adding a threshold needs a bit of thought because it depends on the full transfer size, the configured clock speed and the processor itself. A simple threshold value doesn't suffice here.

benpicco commented:
I can confirm this is working on a `samd51` board (I'll try the `same54-xpro` with the loopback test later).

Ping round-trips with `at86rf215` got worse (from 4.77ms to 6.11ms) but that can be fixed by a follow-up PR that introduces some heuristics for DMA thresholds at different frequencies.

MrKevinWeiss commented:
> @MrKevinWeiss Do you have time somewhere these weeks to test this on the HIL platform?

I suppose but I only have the `samr21-xpro` in there that this PR is related to. For the sake of not breaking anything I will do that.

https://ci.riot-os.org/hil/blue/organizations/jenkins/HIL-Single/detail/HIL-Single/21/pipeline

benpicco commented:
Feel free to add

```patch
diff --git a/boards/same54-xpro/include/periph_conf.h b/boards/same54-xpro/include/periph_conf.h
index bb6ca0c3f0..16e0c7d519 100644
--- a/boards/same54-xpro/include/periph_conf.h
+++ b/boards/same54-xpro/include/periph_conf.h
@@ -164,7 +164,10 @@ static const spi_conf_t spi_config[] = {
         .miso_pad = SPI_PAD_MISO_3,
         .mosi_pad = SPI_PAD_MOSI_0_SCK_1,
         .gclk_src = SAM0_GCLK_48MHZ,
-
+#ifdef MODULE_PERIPH_DMA
+        .tx_trigger = SERCOM4_DMAC_ID_TX,
+        .rx_trigger = SERCOM4_DMAC_ID_RX,
+#endif
     },
     {    /* EXT2, EXT3 */
         .dev      = &(SERCOM6->SPI),
@@ -177,6 +180,10 @@ static const spi_conf_t spi_config[] = {
         .miso_pad = SPI_PAD_MISO_3,
         .mosi_pad = SPI_PAD_MOSI_0_SCK_1,
         .gclk_src = SAM0_GCLK_48MHZ,
+#ifdef MODULE_PERIPH_DMA
+        .tx_trigger = SERCOM6_DMAC_ID_TX,
+        .rx_trigger = SERCOM6_DMAC_ID_RX,
+#endif
     }
 };
```

bergzand commented:
Rebased on #14260 to include the fixes for saml21 and saml1x compatibility. I moved the extra descriptors for the spi registry transfer from the stack to the static allocations. They are conditionally, based on the CPU family, located in the backup sram (LP SRAM). This is mandatory on the SAML platforms.

TL;DR: Now compatible with the SAML21 and SAML1X

bergzand commented:
Added config and compile tested all atsam boards except for the sam3x8e, which seems to be completely incompatible.

benpicco commented:
This fails to compile now

```
/usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/bin/ld: /home/benpicco/dev/RIOT/tests/periph_spi/bin/samr21-xpro/application_tests_periph_spi/main.o: in function `cmd_init':
/home/benpicco/dev/RIOT/tests/periph_spi/main.c:217: undefined reference to `spi_acquire'
```

bergzand commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -75,15 +99,31 @@ void spi_init(spi_t bus)
     while ((dev(bus)->CTRLA.reg & SERCOM_SPI_CTRLA_SWRST) ||
            (dev(bus)->SYNCBUSY.reg & SERCOM_SPI_SYNCBUSY_SWRST)) {}
 
+    dev(bus)->DBGCTRL.bit.DBGSTOP = 1;
```
Leftover from debugging, should be removed

bergzand commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -94,7 +134,7 @@ void spi_init_pins(spi_t bus)
     gpio_init_mux(spi_config[bus].miso_pin, spi_config[bus].miso_mux);
     gpio_init_mux(spi_config[bus].mosi_pin, spi_config[bus].mosi_mux);
-    /* clk_pin will be muxed during acquire / release */
+    gpio_init_mux(spi_config[bus].clk_pin, spi_config[bus].clk_mux);
```
Rebase gone wrong?

bergzand commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -141,18 +181,11 @@ int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
 
-    /* mux clk_pin to SPI peripheral */
-    gpio_init_mux(spi_config[bus].clk_pin, spi_config[bus].clk_mux);
-
```
Same as above

bergzand commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -141,18 +181,11 @@ int spi_acquire(spi_t bus, spi_cs_t cs, spi_mode_t mode, spi_clk_t clk)
-    /* Demux clk_pin back to GPIO_OUT function. Otherwise it will get HIGH-Z
-     * and lead to unexpected current draw by SPI salves. */
-    gpio_disable_mux(spi_config[bus].clk_pin);
-
```
And another one

bergzand commented:
> This fails to compile now

Note that the current [Makefile](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_spi/Makefile#L4) does a `FEATURES_REQUIRED = periph_spi` and not `FEATURES_REQUIRED += periph_spi`, blindly copying that line and modifying it to `periph_dma` will override the required features. (been there, done that.)

benpicco commented:
> > This fails to compile now
> 
> Note that the current [Makefile](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_spi/Makefile#L4) does a `FEATURES_REQUIRED = periph_spi` and not `FEATURES_REQUIRED += periph_spi`, blindly copying that line and modifying it to `periph_dma` will override the required features. (been there, done that.)

This is exactly what happened :laughing:

benpicco commented:
Tested on `saml10-xpro`, works fine.

benpicco commented:
Needs a rebase.

benpicco commented:
Tested on all but `saml21`, works fine.
Please add a test and fix the merge, should be good.

I'd be interesting to run this on the `sam21` though since there things are a bit different with DMA descriptors in LP RAM - maybe @jue89 can give it a try on the `samr30-xpro`?

bergzand commented:
samr30-xpro (on the iot-lab).

<details>
  <summary>without dma</summary>

```
ssh -t zandberg@saclay.iot-lab.info 'socat - tcp:samr30-1.saclay.iot-lab.info:20000'
main(): This is RIOT! (Version: 2020.07-devel-1122-g7c441-pr/sam0_common/spi_dma)
Manual SPI peripheral driver test
Refer to the README.md file for more information.

There are 2 SPI devices configured for your platform.
> init 1 0 4 1 1
init 1 0 4 1 1
SPI_DEV(1) initialized: mode: 0, clk: 4, cs_port: 1, cs_pin: 1
> bench
bench
### Running some benchmarks, all values in [us] ###
### Test                                Transfer time   user time

 1 - write 1000 times 1 byte:                   3911    3924
 2 - write 1000 times 2 byte:                   5842    5855
 3 - write 1000 times 100 byte:         174169  174182
 4 - write 1000 times 1 byte to register:       8501    8514
 5 - write 1000 times 2 byte to register:       10059   10072
 6 - write 1000 times 100 byte to register:     179000  179012
 7 - read 1000 times 2 byte:                    5303    5315
 8 - read 1000 times 100 byte:          174553  174566
 9 - read 1000 times 2 byte from register:      10266   10279
10 - read 1000 times 100 byte from register:    179529  179541
11 - transfer 1000 times 2 byte:                5282    5295
12 - transfer 1000 times 100 byte:              176248  176260
13 - transfer 1000 times 2 byte to register:    10120   10133
14 - transfer 1000 times 100 byte to register:181800    181812
15 - acquire/release 1000 times:                16942   16955
-- - SUM:                                       1141525 1141715

### All runs complete ###
```
</details>

<details>
  <summary>With DMA</summary>

```
ssh -t zandberg@saclay.iot-lab.info 'socat - tcp:samr30-1.saclay.iot-lab.info:20000'
main(): This is RIOT! (Version: 2020.07-devel-1122-g7c441-pr/sam0_common/spi_dma)
Manual SPI peripheral driver test
Refer to the README.md file for more information.

There are 2 SPI devices configured for your platform.
> init 0 0 4 1 1
init 0 0 4 1 1
SPI_DEV(0) initialized: mode: 0, clk: 4, cs_port: 1, cs_pin: 1
> bench
bench
### Running some benchmarks, all values in [us] ###
### Test                                Transfer time   user time

 1 - write 1000 times 1 byte:                   24848   24861
 2 - write 1000 times 2 byte:                   24723   24736
 3 - write 1000 times 100 byte:         105023  28048
 4 - write 1000 times 1 byte to register:       30561   30574
 5 - write 1000 times 2 byte to register:       29961   29974
 6 - write 1000 times 100 byte to register:     110225  32764
 7 - read 1000 times 2 byte:                    24578   24591
 8 - read 1000 times 100 byte:          105251  28291
 9 - read 1000 times 2 byte from register:      30063   30076
10 - read 1000 times 100 byte from register:    110324  32895
11 - transfer 1000 times 2 byte:                24911   24924
12 - transfer 1000 times 100 byte:              105264  28543
13 - transfer 1000 times 2 byte to register:    29961   29974
14 - transfer 1000 times 100 byte to register:110355    32891
15 - acquire/release 1000 times:                16292   16305
-- - SUM:                                       882340  419447

### All runs complete ###
```
</details>

bergzand commented:
> Please add a test

I'll open a separate pull request for a generic SPI with DMA test, inheriting everything from the current SPI test

benpicco commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -164,28 +203,114 @@ void spi_release(spi_t bus)
+
+    if (_use_dma(bus)) {
+        /* The DMA promises not to modify the const out data */
+        _dma_transfer_regs(bus, reg, (uint8_t*)out, (uint8_t*)in, len);
```
No need to cast `in`

benpicco replied:
And now also no need to cast `out` anymore :)

bergzand replied:
Fixed

benpicco commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -164,28 +203,114 @@ void spi_release(spi_t bus)
+    }
+
+    if (_use_dma(bus)) {
+        /* The DMA promises not to modify the const out data */
```
Why not *actually* promise it and make the in `src` parameter `const`

bergzand replied:
Done in #14279

benpicco commented:
Needs a rebase.

bergzand commented:
rebased

bergzand commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -164,28 +203,114 @@ void spi_release(spi_t bus)
+    dma_prepare_dst(_dma_state[bus].rx_dma, in_addr, len, in ? true : false);
+    dma_prepare_src(_dma_state[bus].tx_dma, (void*)out_addr, len,
+                    out ? true : false);
+    _dma_launch(bus);
```
Sure, changed

benpicco commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -82,8 +106,23 @@ void spi_init(spi_t bus)
+        dma_setup(_dma_state[bus].rx_dma,
+                  spi_config[bus].rx_trigger, 1, true);
+        dma_prepare(_dma_state[bus].rx_dma, DMAC_BTCTRL_BEATSIZE_BYTE_Val,
+                    (void*)&dev(bus)->DATA.reg, NULL, 1, 0);
```
You can drop the cast.

bergzand replied:
Unfortunately not, this cast removes the `volatile` qualifier from `DATA.reg`. It doesn't compile without the cast:
```
/home/koen/dev/RIOT-dma/cpu/sam0_common/periph/spi.c: In function 'spi_init':
/home/koen/dev/RIOT-dma/cpu/sam0_common/periph/spi.c:118:21: error: passing argument 3 of 'dma_prepare' discards 'volatile' qualifier from pointer target type [-Werror=discarded-qualifiers]
  118 |                     &dev(bus)->DATA.reg, NULL, 1, 0);
      |                     ^~~~~~~~~~~~~~~~~~~
In file included from /home/koen/dev/RIOT-dma/cpu/samd21/include/periph_cpu.h:25,
                 from /home/koen/dev/RIOT-dma/drivers/include/periph/spi.h:74,
                 from /home/koen/dev/RIOT-dma/cpu/sam0_common/periph/spi.c:31:
/home/koen/dev/RIOT-dma/cpu/sam0_common/include/periph_cpu_common.h:769:56: note: expected 'const void *' but argument is of type 'volatile uint32_t *' {aka 'volatile long unsigned int *'}
  769 | void dma_prepare(dma_t dma, uint8_t width, const void *src, void *dst,
      |                                            ~~~~~~~~~~~~^~~
```

benpicco commented on cpu/sam0_common/periph/spi.c:
```diff
@@ -164,28 +203,115 @@ void spi_release(spi_t bus)
+    const uint8_t *out_addr = out ? out + len : &tmp;
+    uint8_t *in_addr = in ? in + len : &tmp;
+    dma_prepare_dst(_dma_state[bus].rx_dma, in_addr, len, in ? true : false);
+    dma_prepare_src(_dma_state[bus].tx_dma, (void*)out_addr, len,
```
here too

bergzand replied:
Removed!

bergzand commented:
Squashed!

benpicco commented on boards/samr30-xpro/include/periph_conf.h:
```diff
@@ -107,6 +111,10 @@ static const spi_conf_t spi_config[] = {
         .miso_pad = SPI_PAD_MISO_0,
         .mosi_pad = SPI_PAD_MOSI_2_SCK_3,
         .gclk_src = SAM0_GCLK_MAIN,
+#ifdef MODULE_PERIPH_DMA
+        .tx_trigger = SERCOM5_DMAC_ID_TX,
+        .rx_trigger = SERCOM5_DMAC_ID_RX,
+#endif
```
That [vendor file](https://github.com/RIOT-OS/RIOT/blob/master/cpu/sam0_common/include/vendor/samr30/include/instance/sercom5.h#L136) is odd.
No DMA for SERCOM5?

bergzand replied:
> That ~~[vendor file](https://github.com/RIOT-OS/RIOT/blob/master/cpu/sam0_common/include/vendor/samr30/include/instance/sercom5.h#L136)~~ **microcontroller** is odd.

![Screenshot_20200614_161725](https://user-images.githubusercontent.com/5160052/84595858-93e88400-ae5a-11ea-999c-d8c6d594794b.png)

> No DMA for SERCOM5?

No DMA for SERCOM5 :cry:

bergzand replied:
Amended the commit to disable DMA for `SERCOM5` on the samr30-xpro

benpicco commented:
btw.: I played around a bit with finding a way to determine the threshold at which DMA should be used.

My theory was that it takes a constant time to prepare a DMA transfer. That `dma_overhead` (in µs) should be independent of the frequency. And while in the DMA case the bytes are transferred at the theoretical max frequency, for polling there should be a `poll_overhead` with every byte.

So to get the threshold bytes `n` we would simply have to set

```
dma_overhead + n/freq = n*poll_overhead + n/freq
n = dma_overhead/poll_overhead
```

However, that is not what we are seeing. Maybe the poll overhead also depends on the frequency?

```
dma_overhead + n/freq = poll_overhead * n/freq
n = freq*dma_overhead/(poll_overhead - 1)
```

That also starts to diverge with larger frequencies.

Maybe it's a bit of both

```
dma_overhead + n/freq = p1 * n/freq + p2 * n
n = freq*dma_overhead/(p1 + p2*freq - 1)
```

So inserted a hack to toggle DMA at run-time and figure out the threshold bytes for various frequencies: 

<details><summary>hack hack hack</summary>

```patch
diff --git a/cpu/sam0_common/periph/spi.c b/cpu/sam0_common/periph/spi.c
index b7dfc1cd85..08eb91602d 100644
--- a/cpu/sam0_common/periph/spi.c
+++ b/cpu/sam0_common/periph/spi.c
@@ -69,11 +69,18 @@ static inline void poweroff(spi_t bus)
     sercom_clk_dis(dev(bus));
 }
 
+static bool dma_enabled = true;
+
+void sam0_spi_set_dma(bool enabled) {
+    dma_enabled = enabled;
+}
+
 static inline bool _use_dma(spi_t bus)
 {
 #ifdef MODULE_PERIPH_DMA
     return (spi_config[bus].tx_trigger != DMA_TRIGGER_DISABLED) &&
-           (spi_config[bus].rx_trigger != DMA_TRIGGER_DISABLED);
+           (spi_config[bus].rx_trigger != DMA_TRIGGER_DISABLED) &&
+            dma_enabled;
 #else
     (void)bus;
     return false;
diff --git a/tests/periph_spi/Makefile b/tests/periph_spi/Makefile
index 029aa5206b..19bb0e33e0 100644
--- a/tests/periph_spi/Makefile
+++ b/tests/periph_spi/Makefile
@@ -2,6 +2,7 @@ BOARD ?= samr21-xpro
 include ../Makefile.tests_common
 
 FEATURES_REQUIRED = periph_spi
+FEATURES_REQUIRED += periph_dma
 
 USEMODULE += xtimer
 USEMODULE += shell
diff --git a/tests/periph_spi/main.c b/tests/periph_spi/main.c
index adb9d7f919..70549f3bb5 100644
--- a/tests/periph_spi/main.c
+++ b/tests/periph_spi/main.c
@@ -30,6 +30,8 @@
 #include "schedstatistics.h"
 #include "thread.h"
 
+#include "macros/units.h"
+
 /**
  * @brief   Default port number used for the CS pin if unassigned
  */
@@ -61,6 +63,8 @@
 static uint8_t bench_wbuf[BENCH_LARGE];
 static uint8_t bench_rbuf[BENCH_LARGE];
 
+void sam0_spi_set_dma(bool enabled);
+
 extern void sched_statistics_cb(kernel_pid_t active_thread, kernel_pid_t next_thread);
 
 /**
@@ -271,6 +275,69 @@ int cmd_transfer(int argc, char **argv)
     return 0;
 }
 
+static uint32_t get_real_freq(spi_clk_t clk)
+{
+    uint32_t f_src = sam0_gclk_freq(spi_config[spiconf.dev].gclk_src);
+    const uint8_t baud = ((f_src + clk) / (2 * clk) - 1);
+    return f_src / (2 * (baud + 1));
+}
+
+static uint32_t do_bench(bool dma, uint32_t spi_freq, size_t tx_size)
+{
+    uint32_t start, stop;
+
+    sam0_spi_set_dma(dma);
+    spi_acquire(spiconf.dev, spiconf.cs, spiconf.mode, spi_freq);
+
+    start = xtimer_now_usec();
+
+    for (int i = 0; i < BENCH_REDOS; i++) {
+        spi_transfer_bytes(spiconf.dev, spiconf.cs, false,
+                           bench_wbuf, NULL, tx_size);
+    }
+
+    stop = xtimer_now_usec();
+
+    spi_release(spiconf.dev);
+
+    return stop - start;
+}
+
+int cmd_bench_dma(int argc, char **argv)
+{
+    (void)argc;
+    (void)argv;
+
+    uint32_t last_freq = 0;
+
+    for (uint32_t spi_freq = KHZ(100); spi_freq <= MHZ(48); spi_freq += KHZ(100)) {
+
+        size_t tx_size;
+        uint32_t time_dma, time_poll;
+
+        uint32_t spi_freq_real = get_real_freq(spi_freq);
+
+        if (spi_freq_real == last_freq) {
+            continue;
+        }
+
+        last_freq = spi_freq_real;
+
+        for (tx_size = 1; tx_size < 100; ++tx_size) {
+            time_dma  = do_bench(true, spi_freq, tx_size);
+            time_poll = do_bench(false, spi_freq, tx_size);
+
+            if (time_dma < time_poll) {
+                break;
+            }
+        }
+
+        printf("\%lu\t%u\n", spi_freq_real, tx_size);
+    }
+
+    return 0;
+}
+
 int cmd_bench(int argc, char **argv)
 {
     (void)argc;
@@ -545,6 +612,7 @@ static const shell_command_t shell_commands[] = {
     { "init", "Setup a particular SPI configuration", cmd_init },
     { "send", "Transfer string to slave", cmd_transfer },
     { "bench", "Runs some benchmarks", cmd_bench },
+    { "bench_dma", "Runs some DMA benchmarks", cmd_bench_dma },
     { NULL, NULL, NULL }
 };
 ```
</details>

Plotting the results gives in *interesting* graph.

![dma_treshold](https://user-images.githubusercontent.com/1301112/84687630-55c49080-af3e-11ea-8105-166ff86628a7.png)

So for low frequencies, the behavior is indeed linear until we hit some saturation.
We could model that with two parameters.
But what's going on with the behavior at 24 MHz on `samr21`? Looks like we get better results with polling up to 66 bytes there. I don't see that on `same54` (and `saml10` just crashes after 4 MHz)

benpicco commented:
Hm using a higher resolution timer indeed leads to different results

![dma_treshold_timer](https://user-images.githubusercontent.com/1301112/84697729-ef943980-af4e-11ea-8dc6-32038084b15c.png)

<details><summary>12 MHz timer instead of xtimer</summary>

```patch
diff --git a/boards/same54-xpro/include/periph_conf.h b/boards/same54-xpro/include/periph_conf.h
index be145dc676..67b4d754e2 100644
--- a/boards/same54-xpro/include/periph_conf.h
+++ b/boards/same54-xpro/include/periph_conf.h
@@ -42,7 +42,7 @@ static const tc32_conf_t timer_config[] = {
         .mclk           = &MCLK->APBAMASK.reg,
         .mclk_mask      = MCLK_APBAMASK_TC0 | MCLK_APBAMASK_TC1,
         .gclk_id        = TC0_GCLK_ID,
-        .gclk_src       = SAM0_GCLK_8MHZ,
+        .gclk_src       = SAM0_GCLK_48MHZ,
         .flags          = TC_CTRLA_MODE_COUNT32,
     },
     {   /* Timer 1 */
@@ -51,7 +51,7 @@ static const tc32_conf_t timer_config[] = {
         .mclk           = &MCLK->APBBMASK.reg,
         .mclk_mask      = MCLK_APBBMASK_TC2 | MCLK_APBBMASK_TC3,
         .gclk_id        = TC2_GCLK_ID,
-        .gclk_src       = SAM0_GCLK_8MHZ,
+        .gclk_src       = SAM0_GCLK_48MHZ,
         .flags          = TC_CTRLA_MODE_COUNT32,
     }
 };
diff --git a/boards/samr21-xpro/include/periph_conf.h b/boards/samr21-xpro/include/periph_conf.h
index f3a0b7233b..2e303d6baf 100644
--- a/boards/samr21-xpro/include/periph_conf.h
+++ b/boards/samr21-xpro/include/periph_conf.h
@@ -115,11 +115,7 @@ static const tc32_conf_t timer_config[] = {
         .irq            = TC4_IRQn,
         .pm_mask        = PM_APBCMASK_TC4 | PM_APBCMASK_TC5,
         .gclk_ctrl      = GCLK_CLKCTRL_ID_TC4_TC5,
-#if CLOCK_USE_PLL || CLOCK_USE_XOSC32_DFLL
-        .gclk_src       = SAM0_GCLK_1MHZ,
-#else
         .gclk_src       = SAM0_GCLK_MAIN,
-#endif
         .flags          = TC_CTRLA_MODE_COUNT32,
     }
 };
diff --git a/tests/periph_spi/Makefile b/tests/periph_spi/Makefile
index 19bb0e33e0..71734db0b3 100644
--- a/tests/periph_spi/Makefile
+++ b/tests/periph_spi/Makefile
@@ -3,10 +3,11 @@ include ../Makefile.tests_common
 
 FEATURES_REQUIRED = periph_spi
 FEATURES_REQUIRED += periph_dma
+FEATURES_REQUIRED += periph_timer
 
-USEMODULE += xtimer
+# USEMODULE += xtimer
 USEMODULE += shell
 USEMODULE += shell_commands
-USEMODULE += schedstatistics
+# USEMODULE += schedstatistics
 
 include $(RIOTBASE)/Makefile.include
diff --git a/tests/periph_spi/main.c b/tests/periph_spi/main.c
index 70549f3bb5..961caa4826 100644
--- a/tests/periph_spi/main.c
+++ b/tests/periph_spi/main.c
@@ -61,7 +61,7 @@
  * @brief   Benchmark buffers
  */
 static uint8_t bench_wbuf[BENCH_LARGE];
-static uint8_t bench_rbuf[BENCH_LARGE];
+// static uint8_t bench_rbuf[BENCH_LARGE];
 
 void sam0_spi_set_dma(bool enabled);
 
@@ -84,6 +84,7 @@ static struct {
  *
  * Increases the number of context switches by one as a side effect
  */
+#if 0
 static void _sched_statistics_trigger(void)
 {
     sched_statistics_cb(sched_active_pid, sched_active_pid);
@@ -102,6 +103,7 @@ static uint32_t _xtimer_diff_usec(xtimer_ticks32_t stop, xtimer_ticks32_t start)
 {
     return xtimer_usec_from_ticks(xtimer_diff(stop, start));
 }
+#endif
 
 void print_bytes(char* title, uint8_t* data, size_t len)
 {
@@ -289,14 +291,14 @@ static uint32_t do_bench(bool dma, uint32_t spi_freq, size_t tx_size)
     sam0_spi_set_dma(dma);
     spi_acquire(spiconf.dev, spiconf.cs, spiconf.mode, spi_freq);
 
-    start = xtimer_now_usec();
+    start = timer_read(1);
 
     for (int i = 0; i < BENCH_REDOS; i++) {
         spi_transfer_bytes(spiconf.dev, spiconf.cs, false,
                            bench_wbuf, NULL, tx_size);
     }
 
-    stop = xtimer_now_usec();
+    stop = timer_read(1);
 
     spi_release(spiconf.dev);
 
@@ -310,6 +312,10 @@ int cmd_bench_dma(int argc, char **argv)
 
     uint32_t last_freq = 0;
 
+    if (timer_init(1, MHZ(12), NULL, NULL)) {
+        puts("timer init failed");
+    }
+
     for (uint32_t spi_freq = KHZ(100); spi_freq <= MHZ(48); spi_freq += KHZ(100)) {
 
         size_t tx_size;
@@ -343,6 +349,7 @@ int cmd_bench(int argc, char **argv)
     (void)argc;
     (void)argv;
 
+#if 0
     uint32_t start, stop;
     xtimer_ticks32_t sched_start, sched_stop;
     uint32_t sched_diff_us;
@@ -604,7 +611,7 @@ int cmd_bench(int argc, char **argv)
 
     spi_release(spiconf.dev);
     puts("\n### All runs complete ###");
-
+#endif
     return 0;
 }
```
</details>

bergzand commented:
> btw.: I played around a bit with finding a way to determine the threshold at which DMA should be used.

When do you consider it beneficial to use DMA?
 1. When using DMA is faster than using blocking operations?
 2. When using DMA blocks the CPU less than when using blocking operations?

On a related note I've noticed that using DMA with one or two bytes to transfer give varying results depending on whether the interrupt hits before the idle thread can get scheduled or not.

benpicco commented:
If the DMA setup time is greater than what it would take polling (e.g. when only transmitting few bytes), there is nothing to gain from DMA.

And ideally the user doesn't have to care since we can determine those threshold values programmatically.

bergzand commented:
> If the DMA setup time is greater than what it would take polling (e.g. when only transmitting few bytes), there is nothing to gain from DMA.

Agreed.

I've measured the transfer times for different transfer lengths myself.:

![fig](https://user-images.githubusercontent.com/5160052/84782129-f458fc00-afe7-11ea-8def-5b4031905099.png)

The measurements without DMA are linear in time with an overhead per transfer and some overhead per byte. (green, orange and blue)
The DMA-based transfers are slightly faster as they have less or no overhead per byte, but a relative high overhead per transfer. (red, purple and brown)

The CPU busy is measured as the time the CPU is busy with preparing and finishing the DMA transfer. (pink).

Per frequency there are roughly three areas. 
- When polling is faster than setting up the DMA: Always use polling
- When a DMA transfer is faster than using polling: Use DMA when the module is enabled.
- When setting up the DMA is faster than polling, but the DMA-based transfer is slower than polling. 

Now what I want to decide on is whether we should enable DMA in the last case. This can save some power by putting the CPU to sleep, at the cost of some SPI throughput.
For example, this means that with transfers at 5 MHz, between 10 - 30 bytes would also use DMA, and thus have a bit of performance impact compared to using polling for this range.

benpicco commented:
> When setting up the DMA is faster than polling, but the DMA-based transfer is slower than polling.

As it looks like there are quite some irregularities, we can't get it 'perfect' either way.
Just considering DMA setup time would probably be the easier solution to implement too.

bergzand commented:
From my measurements on the samr21-xpro I can model the time required with:
![fig](https://latex.codecogs.com/gif.latex?t_%7Bpolling%7D%20%3D%20t_%7Boverhead%7D%20&plus;%20n_%7Bbytes%7D%20%5Ccdot%20%28overhead_1%20&plus;%20%5Cfrac%7B8%20&plus;%20overhead_2%7D%7Bf_%7Bclk%7D%7D%29)

On the samr21-xpro, t_overhead = 2.54 microsecond, overhead_1 = 0.8 microsecond, overhead_2 = 0.841 * 10 - 9.

I think we should be able to scale and rework the numbers a bit for a fast non-float implementation.

I haven't verified whether these numbers are close enough for the saml21 and I'm curious if the same equation can be fitted to the same54-xpro
